util/oidc/oidc.go

```diff
@@ -149,31 +136,59 @@ func (a *ClientApp) oauth2Config(scopes []string) (*oauth2.Config, error) {
 }
 
 // generateAppState creates an app state nonce
-func (a *ClientApp) generateAppState(returnURL string) string {
+func (a *ClientApp) generateAppState(returnURL string, w http.ResponseWriter) (string, error) {
 	randStr := rand.RandString(10)
 	if returnURL == "" {
 		returnURL = a.baseHRef
 	}
-	err := a.cache.SetOIDCState(randStr, &OIDCState{ReturnURL: returnURL})
-	if err != nil {
-		// This should never happen with the in-memory cache
-		log.Errorf("Failed to set app state: %v", err)
+	cookieValue := fmt.Sprintf("%s:%s", randStr, returnURL)
+	if encrypted, err := crypto.Encrypt([]byte(cookieValue), string(a.settings.ServerSignature)); err != nil {
+		return "", err
+	} else {
+		cookieValue = hex.EncodeToString(encrypted)
 	}
-	return randStr
+
+	http.SetCookie(w, &http.Cookie{
+		Name:     common.StateCookieName,
+		Value:    cookieValue,
+		Expires:  time.Now().Add(3 * time.Minute),
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+		Secure:   a.secureCookie,
+	})
+	return randStr, nil
 }
 
-func (a *ClientApp) verifyAppState(state string) (*OIDCState, error) {
-	res, err := a.cache.GetOIDCState(state)
+func (a *ClientApp) verifyAppState(r *http.Request, w http.ResponseWriter, state string) (string, error) {
+	c, err := r.Cookie(common.StateCookieName)
 	if err != nil {
-		if err == appstatecache.ErrCacheMiss {
-			return nil, fmt.Errorf("unknown app state %s", state)
-		} else {
-			return nil, fmt.Errorf("failed to verify app state %s: %v", state, err)
-		}
+		return "", err
 	}
-
-	_ = a.cache.SetOIDCState(state, nil)
-	return res, nil
+	val, err := hex.DecodeString(c.Value)
+	if err != nil {
+		return "", err
+	}
+	val, err = crypto.Decrypt(val, string(a.settings.ServerSignature))
+	if err != nil {
+		return "", err
+	}
+	cookieVal := string(val)
+	returnURL := ""
+	parts := strings.SplitN(cookieVal, ":", 2)
+	if len(parts) > 1 {
+		returnURL = parts[1]
+	}
```

Comment:
```suggestion
	if len(parts) == 2 {
		returnURL = parts[1]
	} else {
		return "", fmt.Errorf("invalid format in '%s' cookie", common.AuthCookieName)
	}
```

It should never happen. But if I were really eccentric and encrypted my own cookie, I could get Argo CD to redirect me to an empty location, which has [uncertain behavior](https://snook.ca/archives/servers/empty_location).

Reply:
Good catch. I think an implicit check should also be for `parts[1] != ""`, because input string `foo:` would also yield in a len of `parts` of 2, but with an empty second slice item, e.g. `if len(parts) == 2 && parts[1] != "" ...`

Reply:
Thank you! Added a check and unit test